### PR TITLE
OP-1185 | Update MedicalInvenotyRow Add Column IS_NEW_LOT

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OP-1185.

### CODE REVIEW

File: src/main/java/org/isf/medicalstock/manager/MovStockInsertingManager.java [hunk 1]
```diff
@@ -418,4 +418,8 @@ private List<Movement> prepareDishargingMovement(Movement movement, boolean chec
 			return dischargeMovement;
 		}
 	}
+
+	public void deleteLot(Lot lot) throws OHServiceException {
+		ioOperations.deletLot(lot);
```
Comment (mwithi) [inline src/main/java/org/isf/medicalstock/manager/MovStockInsertingManager.java]:
```suggestion
		ioOperations.deleteLot(lot);
```

File: src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java [hunk 2]
```diff
@@ -607,4 +607,8 @@ public long countAllActiveMovements() {
 		return this.movRepository.countAllActiveMovements();
 	}
 
+	public void deletLot(Lot lot) throws OHServiceException {
```
Comment (mwithi) [inline src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java]:
```suggestion
	public void deleteLot(Lot lot) throws OHServiceException {
```

File: src/main/java/org/isf/medicalinventory/model/MedicalInventoryRow.java [hunk 3]
```diff
@@ -144,6 +147,15 @@ public void setRealQty(double realQty) {
 		this.realQty = realQty;
 	}
 	
+	public boolean isNewLot() {
+		return isNewLot;
+	}
+
+
```
Comment (mwithi) [inline src/main/java/org/isf/medicalinventory/model/MedicalInventoryRow.java:154]:
```suggestion

```

File: src/main/java/org/isf/medicalinventory/model/MedicalInventoryRow.java [hunk 4]
```diff
@@ -144,6 +147,15 @@ public void setRealQty(double realQty) {
 		this.realQty = realQty;
 	}
 	
+	public boolean isNewLot() {
+		return isNewLot;
+	}
+
+	
+	public void setNewLot(boolean isNewLot) {
```
Comment (mwithi) [inline src/main/java/org/isf/medicalinventory/model/MedicalInventoryRow.java]:
```suggestion
	}

	public void setNewLot(boolean isNewLot) {
```

File: sql/step_a106_add_movtype_supplier_ward.sql [hunk 5]
```diff
@@ -0,0 +1,4 @@
+INSERT INTO `oh_medicaldsrstockmovtype`(`MMVT_ID_A`, `MMVT_DESC`, `MMVT_TYPE`, `MMVT_CREATED_BY`, `MMVT_CREATED_DATE`, `MMVT_LAST_MODIFIED_BY`, `MMVT_LAST_MODIFIED_DATE`, `MMVT_ACTIVE`) VALUES ('inventory+','Inventory+','+',NULL,NULL,NULL,NULL,1);
+INSERT INTO `oh_medicaldsrstockmovtype`(`MMVT_ID_A`, `MMVT_DESC`, `MMVT_TYPE`, `MMVT_CREATED_BY`, `MMVT_CREATED_DATE`, `MMVT_LAST_MODIFIED_BY`, `MMVT_LAST_MODIFIED_DATE`, `MMVT_ACTIVE`) VALUES ('inventory-','Inventory-','-',NULL,NULL,NULL,NULL,1);
+INSERT INTO `oh_supplier` (`SUP_ID`, `SUP_NAME`, `SUP_ADDRESS`, `SUP_TAXCODE`, `SUP_PHONE`, `SUP_FAX`, `SUP_EMAIL`, `SUP_NOTE`, `SUP_DELETED`, `SUP_CREATED_BY`, `SUP_CREATED_DATE`, `SUP_LAST_MODIFIED_BY`, `SUP_LAST_MODIFIED_DATE`, `SUP_ACTIVE`) VALUES (4'INVENTORY',NULL,NULL,NULL,NULL,NULL,NULL,'N','admin','2024-06-06 10:07:52','admin','2024-06-06 10:14:57',1);
```
Comment (mwithi) [inline sql/step_a106_add_movtype_supplier_ward.sql]:
```suggestion
INSERT INTO `oh_supplier` (`SUP_ID`, `SUP_NAME`, `SUP_ADDRESS`, `SUP_TAXCODE`, `SUP_PHONE`, `SUP_FAX`, `SUP_EMAIL`, `SUP_NOTE`, `SUP_DELETED`, `SUP_CREATED_BY`, `SUP_CREATED_DATE`, `SUP_LAST_MODIFIED_BY`, `SUP_LAST_MODIFIED_DATE`, `SUP_ACTIVE`) VALUES (4,'INVENTORY',NULL,NULL,NULL,NULL,NULL,NULL,'N','admin','2024-06-06 10:07:52','admin','2024-06-06 10:14:57',1);
```

File: src/main/java/org/isf/medicalinventory/model/MedicalInventory.java [hunk 6]
```diff
@@ -23,6 +23,9 @@
 
 import java.time.LocalDateTime;
 
+import org.isf.utils.db.Auditable;
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/model/MedicalInventory.java]: why change the ordering of the imports; the original order was correct.

See the `.ide-settings` folder for the project defaults.

File: src/main/java/org/isf/medicalinventory/model/MedicalInventory.java [hunk 7]
```diff
@@ -156,4 +168,44 @@ public int getLock() {
...
+	
+	public String getDestination() {
+		return destination;
+	}
+
+	
+	public void setDestination(String destination) {
+		this.destination = destination;
+	}
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/model/MedicalInventory.java:225]: why are there no new tests for all this new code?

File: src/main/java/org/isf/medicalinventory/manager/MedicalInventoryRowManager.java [hunk 8]
```diff
@@ -22,19 +22,28 @@
...
 import org.isf.utils.exception.OHServiceException;
+import org.isf.utils.exception.model.OHExceptionMessage;
 import org.springframework.stereotype.Component;
 
+import jakarta.transaction.Transactional;
+
 @Component
 public class MedicalInventoryRowManager {
 
 	private MedicalInventoryRowIoOperation iOoperation;
```
Comment (mwithi) [inline src/main/java/org/isf/medicalinventory/manager/MedicalInventoryRowManager.java]: Refactor

```suggestion
	private MedicalInventoryRowIoOperation ioOperation;
```

File: LOG_DEST/openhospital.log
Comment (mwithi) [inline LOG_DEST/openhospital.log:1]: please remove this file

File: LOG_DEST/openhospital.log.2024-06-17
Comment (mwithi) [inline LOG_DEST/openhospital.log.2024-06-17:1]: please remove this file

File: src/main/java/org/isf/medicalinventory/model/MedicalInventory.java [hunk 9]
```diff
@@ -156,4 +183,44 @@ public int getLock() {
...
+
+	
+	public String getDischangeType() {
+		return dischangeType;
+	}
+
+	
+	public void setDischangeType(String dischangeType) {
+		this.dischangeType = dischangeType;
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/medicalinventory/model/MedicalInventory.java]: Refactor...

```suggestion
	public String getChargeType() {
		return chargeType;
	}

	
	public void setChargeType(String chargeType) {
		this.chargeType = chargeType;
	}

	
	public String getDischargeType() {
		return dischargeType;
	}

	
	public void setDischargeType(String dischargeType) {
		this.dischargeType = dischargeType;
	}
```

File: src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java [hunk 10]
```diff
@@ -577,6 +580,54 @@ public List<Movement> getMovementForPrint(
...
 	}
+	
+	/**
+	 * Retrieves all the lot referred to the specified {@link Medical}, expiring first on top
+	 * 
+	 * @param medical the medical.
+	 * @return a list of {@link Lot}.
+	 * @throws OHServiceException if an error occurs retrieving the lot list.
+	 */
+	public List<Lot> getAllLotsByMedical(Medical medical) throws OHServiceException {
```
Comment (mwithi) [inline src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java]: Why to repeat a full method with only one line of code different? use overloading and DRY principle

Comment (ArnaudFonzam) [reply src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java]: I have removed it and update method `getLotsByMedical(Medical medical)` by this `getLotsByMedical(Medical medical, boolean removeEmpty)`

File: src/main/java/org/isf/medicalinventory/manager/MedicalInventoryRowManager.java [hunk 11]
```diff
@@ -56,7 +65,17 @@ public MedicalInventoryRow newMedicalInventoryRow(MedicalInventoryRow medicalInv
...
+		if (medInvRow.isPresent()) {
+			MedicalInventoryRow medInvR = medInvRow.get();
+			medInvR.setLot(medicalInventoryRow.getLot());
+			medInvR.setRealqty(medicalInventoryRow.getRealQty());
+			if (medicalInventoryRow.isNewLot()) {
+				medInvR.setNewLot(true);
+			}
+			return ioOperation.updateMedicalInventoryRow(medInvR);
+		}
+		throw new OHDataValidationException(new OHExceptionMessage(MessageBundle.getMessage("angal.inventoryrow.notfound.msg")));
```
Comment (mwithi) [inline src/main/java/org/isf/medicalinventory/manager/MedicalInventoryRowManager.java]: Why is this needed? Updating an existing entity (same @Id) it should just go through without retrieving it and resetting its properties... or maybe I'm missing something?

Comment (ArnaudFonzam) [reply src/main/java/org/isf/medicalinventory/manager/MedicalInventoryRowManager.java]: I will add validation method to avoid that check

File: src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java [hunk 12]
```diff
@@ -136,6 +143,8 @@ public List<MedicalInventory> getMedicalInventory() throws OHServiceException {
 	 * @throws OHServiceException
 	 */
 	public List<MedicalInventory> getMedicalInventoryByParams(LocalDateTime dateFrom, LocalDateTime dateTo, String status, String type) throws OHServiceException {
+		dateFrom = LocalDateTime.of(dateFrom.toLocalDate(), LocalTime.MIN);
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java]: why not use TimerTools, `getBeginningOfDay()` and `getBeginningOfNextDay()`?

File: src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java [hunk 13]
```diff
@@ -152,6 +161,54 @@ public List<MedicalInventory> getMedicalInventoryByParams(LocalDateTime dateFrom
...
+	 */
+	public MedicalInventory getInventoryByReference(String reference) throws OHServiceException {
+		return ioOperations.getInventoryByReference(reference);
+	}
+	
+	/**
+	 * Verify if the object is valid for CRUD and return a list of errors, if any.
+	 *
+	 * @param medInventory
+	 * @throws OHDataValidationException
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java:195]: The method signature has `OHServiceException` and the code throws `OHDataValidationException`

File: src/main/java/org/isf/medicalinventory/manager/MedicalInventoryRowManager.java [hunk 14]
```diff
@@ -76,6 +88,70 @@ public void deleteMedicalInventoryRow(MedicalInventoryRow medicalInventoryRow) t
...
+		return null;
+	}
+	
+	/**
+	 * Verify if the object is valid for CRUD and return a list of errors, if any.
+	 *
+	 * @param medInventoryRow
+	 * @throws OHDataValidationException
+	 */
+	private void validationMedicalInventoryRow(MedicalInventoryRow medicalInventoryRow) throws OHDataValidationException {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/manager/MedicalInventoryRowManager.java]: Doesn't other code in the project use `validate` for the first part of the method name instead of `validation`?

File: src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java [hunk 15]
```diff
@@ -152,6 +161,54 @@ public List<MedicalInventory> getMedicalInventoryByParams(LocalDateTime dateFrom
...
+	 */
+	private void validationMedicalInventory(MedicalInventory medInventory) throws OHServiceException {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java]: Doesn't other code in the project use `validate` for the first part of the method name instead of `validation`?

File: src/main/java/org/isf/medicalinventory/model/MedicalInventory.java [hunk 16]
```diff
@@ -74,7 +74,19 @@ public class MedicalInventory extends Auditable<String> {
...
 	private String ward;
+	
+	@Column(name = "MINVT_CHARGE_TYPE")
+	private String chargeType;
+	
+	@Column(name = "MINVT_DISCHARGE_TYPE")
+	private String dischargeType;
+	
+	@Column(name = "MINVT_SUPPLIER")
+	private Integer supplier;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/model/MedicalInventory.java:85]: why is this an `Integer`?  Can it be null?  If not then I think it should be an `int`

File: src/main/java/org/isf/medicalinventory/service/MedicalInventoryIoOperation.java [hunk 17]
```diff
@@ -171,4 +172,30 @@ public Page<MedicalInventory> getMedicalInventoryByParamsPageable(LocalDateTime
...
+	 */
+	public MedicalInventory getInventoryByReference(String  reference) throws OHServiceException {
+		return repository.findByReference(reference);
+	}
+
+	/**
+	 * Fetch {@link MedicalInventory} with param.
+	 * 
+	 * @param inventoryId - the {@link MedicalInventory} id.
+	 * @return {@link MedicalInventory}. It could be {@code empty}.
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/service/MedicalInventoryIoOperation.java]: Shouldn't this say it could be `null`?

File: src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java [hunk 18]
```diff
@@ -311,6 +311,9 @@ public Lot updateLot(Lot lot) throws OHServiceException {
 	 */
 	// TODO: verify why lotCode and medical params are needed
 	public Lot storeLot(String lotCode, Lot lot, Medical medical) throws OHServiceException {
+		if (lotCode.equals("")) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java]: What happens if `lotCode` is `null`?

File: src/test/java/org/isf/medicalstock/Tests.java [hunk 19]
```diff
@@ -229,7 +229,7 @@ void testIoGetLotsByMedical(boolean in, boolean out, boolean toward) throws Exce
 		int code = setupTestMovement(false);
 		Movement foundMovement = movementIoOperationRepository.findById(code).orElse(null);
 		assertThat(foundMovement).isNotNull();
-		List<Lot> lots = medicalStockIoOperation.getLotsByMedical(foundMovement.getMedical());
+		List<Lot> lots = medicalStockIoOperation.getLotsByMedical(foundMovement.getMedical(), true);
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/medicalstock/Tests.java:232]: All the tests appear to use the `true` parameter; are there any that use `false`?

File: src/main/java/org/isf/medicalinventory/model/MedicalInventory.java (hunk 16 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/medicalinventory/model/MedicalInventory.java:85]: yes It can be null

File: src/test/java/org/isf/medicalstock/Tests.java (hunk 19 above)
Comment (ArnaudFonzam) [reply src/test/java/org/isf/medicalstock/Tests.java:232]: yes I added

File: src/main/java/org/isf/medicalinventory/service/MedicalInventoryIoOperation.java (hunk 17 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/medicalinventory/service/MedicalInventoryIoOperation.java]: yes I update its

File: src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java (hunk 18 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/medicalstock/service/MedicalStockIoOperations.java]: I added also test for null value

File: src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java [hunk 20]
```diff
@@ -152,6 +161,54 @@ public List<MedicalInventory> getMedicalInventoryByParams(LocalDateTime dateFrom
...
+		return ioOperations.getInventoryByReference(reference);
+	}
+	
+	/**
+	 * Verify if the object is valid for CRUD and return a list of errors, if any.
+	 *
+	 * @param medInventory
+	 * @throws OHServiceException
+	 */
+	private void validateMedicalInventory(MedicalInventory medInventory) throws OHServiceException {
```
Comment (mwithi) [inline src/main/java/org/isf/medicalinventory/manager/MedicalInventoryManager.java]:
```suggestion
	 * @throws OHDataValidationException
	 */
	private void validateMedicalInventory(MedicalInventory medInventory) throws OHDataValidationException {
```

File: src/main/java/org/isf/medicalstock/manager/MovStockInsertingManager.java [hunk 21]
```diff
@@ -237,17 +237,18 @@ private boolean isAutomaticLotOut() {
 	}
 
 	/**
-	 * Retrieves all the {@link Lot} associated to the specified {@link Medical}, expiring first on top
+	 * Retrieves all the {@link Lot} associated to the specified {@link Medical}, expiring first on top with zero quantities will be stripped out if removeEmpty is set to true.
```
Comment (mwithi) [inline src/main/java/org/isf/medicalstock/manager/MovStockInsertingManager.java]:
```suggestion
	 * Retrieves all the {@link Lot} associated to the specified {@link Medical}, expiring first on top, zero quantities will be stripped out if {@code removeEmtpy} is set to true.
```

File: src/main/java/org/isf/medicalstock/manager/MovStockInsertingManager.java
Comment (mwithi) [inline src/main/java/org/isf/medicalstock/manager/MovStockInsertingManager.java:1]: Please add this method to avoid issues in API

```
	/**
	 * Retrieves all the {@link Lot} associated to the specified {@link Medical}, expiring first on top, zero quantities will be stripped out.
	 * 
	 * @param medical the medical.
	 * @return the list of retrieved {@link Lot}s.
	 * @throws OHServiceException
	 */
	public List<Lot> getLotByMedical(Medical medical) throws OHServiceException {
		return ioOperations.getLotsByMedical(medical, true);
	}
```

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!
